### ISSUE TITLE
Sync device counts should be consistent during a login flow

### ISSUE THREAD

**Issue by 6a68** [mozilla/fxa#6578]
While signing in to sync, the desktop browser makes a number of auth-server requests:

```
/account/login
/oauth/token <-- issues 'profile' token
/verify
/account/keys
/account/device <-- device count is incremented here
/oauth/token <-- issues 'oldsync' token
/account/scoped-key-data
```

(If interested, you can test this yourself by going to about:config, creating a boolean pref `identity.fxaccounts.allowHttp` and setting it to true, then filling in `identity.fxaccounts.autoconfig` with `http://localhost:3030`, then restarting. You'll now be able to sign in to local FxA from the browser toolbar FxA button, and you can watch the auth-server logs or network tab in devtools.)

Some events are logged in the course of this flow that include device info about the account, a little JS object of the form

```
{ sync_device_count, sync_active_devices_day, sync_active_devices_week, sync_active_devices_month }
```

Based on when in this flow the events are logged, they will carry the wrong device count.

Events logged before the device is incremented include:
* `fxa_email - sent`
* `fxa_activity - oauth_access_token_created` for the `profile` scope (this is why sync device counts are off by one)
* `fxa_activity - cert_signed` (not because we sign a cert, but to trigger `fxa_login - complete`, see #6347)
* `fxa_login - complete`

Events logged after the device count is incremented:
* `fxa_activity - oauth_access_token_created` for the `oldsync` scope (but missing the `event_properties: {service: sync}` on the event, which seems to exclude it from the sync device counts rollup)
* `fxa_activity - cert_signed` (not yet sure why this event is fired twice, but probably related to #6347)

We should rethink how we are recording these numbers, since the current approach isn't working. 

-----

Interestingly, one of the first things that happens in a sync login flow is [logging a "notify attached services" event](https://github.com/mozilla/fxa/blob/main/packages/fxa-auth-server/lib/routes/utils/signin.js#L306-L314), which includes a `deviceCount` field based on the session count, which seems to be correct in local testing.

Options:

* We could get rid of the `oauth_access_token_created` event, and instead modify `fxa_login - complete` to fire _after_ the device count is incremented, and base our sync counts on this event. (I think this is probably the right fix, but altering the order of events might cause some other bug to emerge.)

* Or, we could just base our device counts on `db.sessions.count` to track our sync metrics? (Not sure this will track sync logins per day/week/month without some more work, and not sure if this number will always be correct)

* Or, should we optimistically increment the device count as soon as login starts, and decrement it if something fails?

**Comment by rfk** [comment]
> We could get rid of the oauth_access_token_created event

I don't think we should get rid of it entirely, it can be used for other purposes than what we're tracking here.

> (I think this is probably the right fix, but altering the order of events might cause some other bug to emerge.)

Changing the order here does seem like the right way to go though, and if some other bugs emerge, well, that's going to be frustrating but ultimately useful... 😬

**Comment by 6a68** [comment]
Thanks to #6581, it looks like the `fxa_login - complete` event is now logged after the sync device count is incremented.

Do we still need the `oauth_access_token_created` event? I'm not sure.

The `fxa_activity - oauth_access_token_created` events for both the profile and oldsync scopes now report `{service: "undefined_oauth"}`, which doesn't seem like an improvement. I wonder if it would make sense to special-case the `service` to `sync` for this event as well as the `fxa_login - complete` event, looking at this section of code: https://github.com/chenba/fxa/blob/4d5cca23adcc907dce43321b2397f095712dc1e0/packages/fxa-auth-server/lib/routes/oauth/index.js#L321-L327

**Comment by chenba** [comment]
A quick before-signing-off-on-Friday comment: the oauth service names rely on a map from the configs.  The reason you get that value is because there's no default map in local env (...pretty sure).  See `oauth.clientIds` in auth-server configs, and grep for `oauth_client_id_map` in the cloudops-deployment repo.

**Comment by data-sync-user** [comment]
➤ Jared Hirsch commented:

I suspect this may be fixed by Barry's #6581 PR, but we won't know until last week's train hits production–which won't be until nearly the end of the sprint. Kicking this to the backlog for now.

**Comment by data-sync-user** [comment]
➤ Alex Davis commented:

[~jhirsch@mozilla.com] I don't know if I'm mixing issues here but the overall device count in Amplitude doesn't seem to be resolved yet:
https://analytics.amplitude.com/mozilla-corp/chart/n2bn398/edit/epnb2io ( https://analytics.amplitude.com/mozilla-corp/chart/n2bn398/edit/epnb2io )

Though, Barry's chart does allow us to track active Sync users again:
https://analytics.amplitude.com/mozilla-corp/chart/new/2w1gqpq ( https://analytics.amplitude.com/mozilla-corp/chart/new/2w1gqpq )

**Comment by 6a68** [comment]
@davismtl Yup, I think I've figured out the root cause for the overall device count bug. The chart I linked shows that, as sync moved to oauth, the number of devices per user per day incorrectly dropped by one, so the old 1 device/day user curve looks like the new 0 device/day user curve. I suspect weirdness in the week/month aggregates is a side effect of this daily aggregation bug.

I suspect this is related to the timing issues I documented in #6578. Looking at that issue, we are now recording `fxa_login - complete` events when the profile token is requested, but before the device has been added to the account, hence it's off by one (in most cases, recording 0 devices where it should be recording 1 device). I suspect that we can instead fire `fxa_login - complete` when the oldsync scope is requested, which happens after the device has been added to the account. Working on a patch now.

**Comment by 6a68** [comment]
I did some Fenix testing this week, and the token/device sequence is slightly different:

(note: these are all POST requests unless indicated otherwise, and note also I've skipped unrelated network traffic)
```
/v1/oauth/token   <-- 'fxa-credentials' grant for token with profile & clients:write scopes
/v1/oauth/authorization <-- swaps the token for an auth code
/v1/oauth/token   <-- auth code grant for access, refresh, & session tokens with profile & oldsync scopes
/v1/account/device  <-- adds the device to the account
/v1/oauth/token   <-- uses refresh token to request oldsync token
GET /v1/account/devices     }
/v1/account/device          }  not 100% sure but seems like the device is just updated here
/v1/oauth/token   <-- uses refresh token to request profile token
```
So, looking at the request ordering here, and building on discussion and work in #6581, if we only trigger the flow complete event in response to a request for the oldsync scope alone (not profile scope, and not both profile and oldsync scopes), I think we will record correct device counts for both mobile and desktop.

The rollup looks at the max count for all pings recorded for a given user ID per day, so technically I guess we don't need to do this to fix sync device counts over in #5143. But still maybe worth enforcing?
